### ISSUE TITLE
Amazon Mechanical Turk (mturk) compatibility

### ISSUE THREAD

**Issue by mpizenberg** [mpizenberg/annotation-app#23]
To annotate a small amount of data, one can just do it on their computer. For this the default interface is supposed to be a perfect match. At the other end of the spectrum, some annotation tasks are just too huge, and need to be crowd-sourced. Amazon Mechanical Turk (mturk) is the perfect service for that. It comes in two sides. The "**requester**" is the one setting a set of tasks. The "**worker**" is the one doing the tasks. Workers are payed by requesters through mturk service.

To be compatible with mturk, a few things need to be changed. First the concept of a "**HIT**" (Human Intelligence Task) should be easily quantified. The simplest way of quantification is **one image <=> one HIT**. So one and only one image should be displayed per HIT. Therefore, there is **no need to display the name of the image on the right sidebar**. Also, the **configuration should be defined by the requester** and not uploadable by a worker.

The interface in mturk can be provided with an html file, which is plugged in an html frame whithin mturk web page. Thus, once the rest is working, using mturk with our interface boils down to **finding the simplest html file compatible with mturk** and our app. Static (`utils.js`, `ports.js`, `elm-pep.js`) and compiled (`Main.js`) should be available at an **https** address to be embeded correctly as scripts tags within the html file. The utils functions are not useful in mturk context so we can just **remove `utils.js`** from the html file script imports.

Submitting a HIT to mturk is done through a form submit. Our html interface is plugged into an already prepared form. Mturk forces us to have at least one `input` tag in our html interface. So things we have to do are: **hiding mturk form** (`#mturk_form { display : none }`), and in the save/export port, instead of saving the Json file on disk, put the Json data into our input `value`, and **programatically `click` on the `#submitButton`**.

To recap, the things to do are:

* [x] Pass config and images in startup flags
* [x] Have a mturk flag changing the top bar buttons view
* [x] Do not show images side bar when only one image loaded
* [x] Create an adapted `ports-mturk.js` file with only necessary functions
* [x] Create an mturk-compatible html file, with an input, no utils, the adapted ports.
* [x] Create a SSL certificate for our website with let's encrypt to be able to serve js files with https

**Comment by mpizenberg** [comment]
SSL certificate turned out to be quite easy :) thanks to the work of the [Electronic Frontier Foundation (EFF)][certbot], and the [Let's Encrypt][encrypt] project. I'm using Nginx on a Centos 7 vps to serve the application, so this was done in three simple steps.

[certbot]: https://certbot.eff.org/
[encrypt]: https://letsencrypt.org/

**Step 1: installing certbot**
```shell
sudo yum install certbot
sudo yum install certbot-nginx
```

**Step 2: launch certbot interactive mode to update nginx config**
```shell
# Just follow the bot questions
sudo certbot --nginx
```

**Step 3: configure automatic certificate renewal once a day**
```shell
# Make sure everything is fine with a dry-run renewal
sudo certbot renew --dry-run

# Add a cron job
sudo crontab -e

# Inside the opened file, add the following line, save and close
*/24 * * * * certbot renew --quiet --no-self-upgrade && systemctl restart nginx
```

Done :tada:

**Comment by mpizenberg** [comment]
Creating an mturk-compatible html file is harder than expected because of mturk templating system. Mturk is replacing variables of the form `${some_var}` in the plugged html by values provided in a CSV file, when "Publishing a batch". I have encountered at least two issues with it:

1. The template system erase one `}` when there are two successive `}`. For example with `const x = { a:'a', b:${some_var} }`. The only way to deal with this is to write those on multiple lines, and to put a `// comment` after the first `}`.
2. It is impossible (or I couldn't manage) to put some Json string inside of the provided CSV batch file, due to the `,` and `"` characters.

As a consequence, we will set the config in the html file instead of as a CSV entry. It's not such an issue since we probably want to keep the same interface for all hits of a same batch.

For now I have it working (with a tiny complication explained later) with the following pair of files.

The html template:

```html
<div>
	<input type="hidden" value="" name="annotation-data" id="annotation-data"/>
	<style>html, body, .style-elements { height: 100% } #mturk_form { display: none }</style>
	<script src="https://annotation-app.pizenberg.fr/elm-pep.js"></script>
	<script src="https://annotation-app.pizenberg.fr/Main.js"></script>
	<script charset="utf-8">
		// Function returning the size of the container element for the app.
		// In our case, the full layout viewport.
		const layoutViewportSize = () => ({
			width: document.documentElement.clientWidth,
			height: document.documentElement.clientHeight
		});
		const containerSize = layoutViewportSize;

		// The image to display
		const img = {
			width: ${img_width},
			height: ${img_height},
			url: "${img_url}"
			// comment to prevent mturk templating system to merge those two '}'
		};

		// Set elm app flags
		const flags = {
			deviceSize: containerSize(),
			mturkMode: true,
			images: [img],
			config: `{
				"classes": [],
				"annotations": [ "point", "bbox", "stroke", "outline", "polygon" ]
			}`
		};

		// Start elm app.
		const app = Elm.Main.fullscreen(flags);
	</script>
	<script src="https://annotation-app.pizenberg.fr/ports-mturk.js"></script>
</div>
```

The CSV batch file:

```csv
img_width,img_height,img_url
571,414,https://mpizenberg.github.io/resources/elm-image-annotations/images/rosa_annotated.jpg
...
width_n,height_n,url_n
```

The tiny complication is that the image is often not displayed. Simply resizing the window is sufficient to trigger display of the image but I will have to figure out why the image does not show up.
